src/Alpacc/RegularExpression.hs

```diff
@@ -0,0 +1,60 @@
+module Alpacc.RegularExpression (regExFromText, pRegEx) where
+
+import Control.Monad.Combinators.Expr
+import Data.Text (Text)
+import Data.Void
+import Text.Megaparsec
+import Text.Megaparsec.Char (space1)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+
+type Parser = Parsec Void Text
+
+data RegEx
+  = Epsilon
+  | Literal Char
+  | Star RegEx
+  | Alter RegEx RegEx
+  | Concat RegEx RegEx
+  deriving (Eq, Show)
+
+space :: Parser ()
+space = Lexer.space space1 (Lexer.skipLineComment "#") empty
+
+lexeme :: Parser a -> Parser a
+lexeme = Lexer.lexeme space
+
+symbol :: Text -> Parser Text
+symbol = Lexer.symbol space
+
+pLiteral :: Parser RegEx
+pLiteral = Literal <$> lexeme (satisfy (`elem` ['a' .. 'z']))
+
+binary :: Text -> (RegEx -> RegEx -> RegEx) -> Operator Parser RegEx
+binary name f = InfixL (f <$ symbol name)
+
+postfix :: Text -> (RegEx -> RegEx) -> Operator Parser RegEx
+postfix name f = Postfix (f <$ symbol name)
+
+operatorTable :: [[Operator Parser RegEx]]
+operatorTable =
+  [ [ postfix "*" Star
+    ],
+    [ binary "" Concat
+    ],
+    [ binary "|" Alter
+    ]
+  ]
```

Comment:
Using these generators for a grammar as simple as regular expressions is a bit of a trap.  I suggest just using multiple levels of nonterminals, like here: https://github.com/athas/scripts/blob/93e1225d8aae43495034daf6115235f32636826e/HsGrep.hs#L42-L61

Reply:
I think I have fixed it, I think I was a little confused when I could not find chainl1 in Megaparsec.